### ISSUE TITLE
[@golevelup/ts-jest & @golevelup/nestjs-rabbitmq]

### ISSUE THREAD

**Issue by calvintwr** [golevelup/nestjs#475]
I have an AppModule file as follows:

```ts
@Module({
    imports: [
        RabbitMQModule.forRoot(RabbitMQModule, {
            exchanges: [
                {
                    name: RABBIT_MQ.EXCHANGE_NAME,
                    type: 'direct',
                },
            ],
            uri: process.env.RABBITMQ_URI,
            connectionInitOptions: { wait: true },
        }),
    ],
})
export class AppModule {}
```

I have tried to mock rabbitmq using ts-jest like this:

```ts
beforeEach(async () => {
        const module: TestingModule = await Test.createTestingModule({
            imports: [
                AppModule
            ],
        })
            .overrideProvider(AmqpConnection)
            .useValue(createMock<AmqpConnection>())
            .compile()
    })
```

This is giving me error:

```
[Nest] 2745  - 24/07/2022, 17:02:54   ERROR [AmqpConnection] Disconnected from RabbitMQ broker (default)
Error: connect ECONNREFUSED 127.0.0.1:5672
```

If i mock the whole rabbitmq module like:

```ts
jest.mock('@golevelup/nestjs-rabbitmq')
```

I will get errors like:

```
Nest cannot create the AppModule instance.
    The module at index [0] of the AppModule "imports" array is undefined.
```

Has anyone successfully mocked RabbitMQ? Please assist if possible.

**Comment by calvintwr** [comment]
## Update

Turns out, this is not a good answer below. Please see https://stackoverflow.com/a/75565940/2755480

## Original post
-------------------------------
I found the answer here:

https://stackoverflow.com/a/70420769/2755480

TLDR;

```ts
import { Module } from '@nestjs/common'

@Module({})
class GenericMockedModule {}

jest.mock('@golevelup/nestjs-rabbitmq', () => {
    const originalModule = jest.requireActual('@golevelup/nestjs-rabbitmq')

    return Object.assign(originalModule, {
        RabbitMQModule: {
            forRoot: jest.fn(() => GenericMockedModule),
        },
        // whichever other interfaces you need to use, you can mock it here.
    })
})

```

**Comment by adifaidz** [comment]
@calvintwr Did you mock `AmqpConnection` inside of the module mocking or you did it like the previous example?

**Comment by ezebrotsky** [comment]
> @calvintwr Did you mock `AmqpConnection` inside of the module mocking or you did it like the previous example?

@adifaidz this question is important, if AmqpConnection is not mocked, I get the following error:
```
Nest can't resolve dependencies of the Service (?). Please make sure that the argument dependency at index [0] is available in the Module context.

    Potential solutions:
    - If dependency is a provider, is it part of the current Module?
    - If dependency is exported from a separate @Module, is that module imported within Module?
      @Module({
        imports: [ /* the Module containing dependency */ ]
      })
```

In the service I have:
```
@Injectable()
export class Service {
  constructor(private readonly amqpConnection: AmqpConnection) {}
```

**Comment by calvintwr** [comment]
@adifaidz @ezebrotsky Write a manual mock and put it in `src/__mocks/@golevelup/nestjs-rabbitmq`:

```js
import { mock } from 'jest-mock-extended'

// create a deeply mocked module
const rmq = jest.createMockFromModule<typeof import('@golevelup/nestjs-rabbitmq')>(
    '@golevelup/nestjs-rabbitmq',
)

// patch the methods that needs a mock interface
rmq.RabbitMQModule.forRoot = jest.fn(() => ({
    module: rmq.RabbitMQModule,
    providers: [
        {
            provide: rmq.AmqpConnection,
            useValue: mock<typeof rmq.AmqpConnection>(),
        },
    ],
    exports: [rmq.AmqpConnection],
}))

module.exports = rmq
```

See my full answer here: https://stackoverflow.com/a/75565940/2755480

**Comment by ezebrotsky** [comment]
@calvintwr thank you! It worked for me!
